calico_containers/adapter/datastore.py

```diff
@@ -746,32 +867,31 @@ def set_endpoint(self, hostname, container_id, endpoint):
         """
         ep_path = ENDPOINT_PATH % {"hostname": hostname,
                                    "container_id": container_id,
-                                   "endpoint_id": endpoint.ep_id}
-        self.etcd_client.write(ep_path, endpoint.to_json())
+                                   "endpoint_id": endpoint.endpoint_id}
+        new_json = endpoint.to_json()
+        self.etcd_client.write(ep_path, new_json)
+        endpoint._original_json = new_json
 
     @handle_errors
-    def update_endpoint(self, hostname, container_id,
-                        old_endpoint, new_endpoint):
+    def update_endpoint(self, endpoint):
         """
-        Update a single endpoint object to the datastore. Fails if the
-        old_endpoint that's passed in doesn't match what's in the datastore.
+        Update a single endpoint object to the datastore.  This assumes the
+        endpoint was originally queried from the datastore and updated.
         Example usage:
-            old_endpoint = datastore.get_endpoint(...)
-            new_endpoint = old_endpoint.copy()
+            endpoint = datastore.get_endpoint(...)
             # modify new endpoint fields
-            datastore.update_endpoint(..., old_endpoint, new_endpoint)
+            datastore.update_endpoint(endpoint)
 
-        :param hostname: The hostname for the Docker hosting this container.
-        :param container_id: The Docker container ID.
-        :param old_endpoint: The existing endpoint to update.
-        :param new_endpoint: The Endpoint to add to the container.
+        :param endpoint: The Endpoint to add to the container.
         """
-        ep_path = ENDPOINT_PATH % {"hostname": hostname,
-                                   "container_id": container_id,
-                                   "endpoint_id": new_endpoint.ep_id}
+        ep_path = ENDPOINT_PATH % {"hostname": endpoint.hostname,
+                                   "container_id": endpoint.workload_id,
```

Comment:
Can we rename the container_id in these string templates to be workload_id ... (if its more that 15 minutes work then just add to the google doc to do list)